include/deal.II/distributed/grid_tools.h

```diff
@@ -0,0 +1,333 @@
+     * objects from locally owned cells to ghost cells on other processors.
+     *
+     * After this call, you will have received data from @p unpack on every
+     * ghost cell as it was given by @p pack on the owning processor.
+     *
+     * @tparam DataType The type of the data to be communicated. It is assumed
+     * to be serializable by boost::serialization.
+     * @tparam MeshType The type of @p mesh.
+     *
+     * @param mesh A variable of a type that satisfies the requirements of the
+     * @ref ConceptMeshType "MeshType concept".
+     * @param pack The function that will be called on each locally owned cell
+     * that needs to be sent.
+     * @param unpack The function that will be called for each ghost cell
+     * with the data imported from the other procs.
+
+     */
+    template <typename MeshType, typename DataType>
+    void
+    exchange_cell_data_to_ghosts (const MeshType &mesh,
+                                  std::function<DataType (const typename MeshType::active_cell_iterator &)> pack,
+                                  std::function<void (const typename MeshType::active_cell_iterator &, const DataType &)> unpack);
+  }
+
+
+}
+
+#ifndef DOXYGEN
+
+namespace parallel
+{
+  namespace GridTools
+  {
+
+    namespace internal
+    {
+      /**
+       * A structure that allows the transfer of data of type T from one processor
+       * to another. It corresponds to a packed buffer that stores a list of
+       * cells and an array of type T.
+       *
+       * The vector @p data is the same size as @p cell_ids.
+       */
+      template <int dim, typename T>
+      struct CellDataTransferBuffer
+      {
+        std::vector<CellId> cell_ids;
+        std::vector<T> data;
+
+        /**
+         * Write the data of this object to a stream for the purpose of
+         * serialization.
+         */
+        template <class Archive>
+        void save (Archive &ar,
+                   const unsigned int /*version*/) const
+        {
+          Assert(cell_ids.size() == data.size(), ExcInternalError());
+          // archive the cellids in an efficient binary format
+          ar &cell_ids.size();
+          for (auto &it : cell_ids)
+            {
+              CellId::binary_type binary_cell_id = it.template to_binary<dim>();
+              ar &binary_cell_id;
+            }
+
+          ar &data;
+        }
+
+        /**
+         * Read the data of this object from a stream for the purpose of
+         * serialization. Throw away the previous content.
+         */
+        template <class Archive>
+        void load (Archive &ar,
+                   const unsigned int /*version*/)
+        {
+          size_t n_cells;
+          ar &n_cells;
+          cell_ids.clear();
+          cell_ids.reserve(n_cells);
+          for (unsigned int c=0; c<n_cells; ++c)
+            {
+              CellId::binary_type value;
+              ar &value;
+              cell_ids.emplace_back(value);
```

Comment:
on second thought, I think `emplace_back (std::move(value))` is even more efficient.